Resuming from Sleep and other connection fixes

gbakeman opened the pull request:
# Socket
- Simplified `ConnectionStatus` boolean property while maintaining null check
- Simplified NUT and Net version properties
- Removed redundant `Socket` object in favor of `TcpClient` for underlying socket management
- Removed unused `SocketDisconnected` event
- Simplify Connection logic to only do the basic work of establishing streams for reading and writing
- Gather version information immediately after successful connection, and within error tolerant try-catch blocks so as to bypass non-critical errors
- Login subroutine refers to instance fields instead of parameters (login info should never change during lifetime of the instance)
- Corrected `LOGIN` command syntax
- Disconnect subroutine is more likely to dispose of all related objects when called
- Support NUT network version 1.3

# UPS Device
- Small tweaks to properties
- Removed ReConnected event - was duplicating calls with the Connected event
- Polling interval is now read and set correctly. No longer writable after instantiation.
- Wrap Login method for socket

# WinNUT
- Remove unused `UpdateMethod` property and field
- Add support for `Suspend` system power mode change to try to prepare WinNUT for suspension (isn't always called before system suspends.)
- Reinforced `Resume` system power mode change to cleanup WinNUT's state if it was left connected while suspending previously
- WinNUT initiates the call to Login to the UPS now (support for accessing UPS without a login as intended by NUT)

## Related
- #145 
- #162

gbakeman committed:
VS automatic changes to .resx files

gbakeman committed:
Support Suspend power mode change

We previously didn't support the Suspend power mode change, which meant WinNUT would be paused in the middle of execution (and with an open socket) if the system goes to sleep. Now WinNUT has the chance to close down operations and connections, which seems to at least help with errors while changing power states. It looks like WinNUT still tries to reconnect too quickly after the system resumes from sleep (before networking is fully restored?) In the interest of reducing notification spam, it may be best to have WinNUT wait a few seconds before reconnecting.

gbakeman committed:
Removing Form-wide manual update variable

Variable was essentially unused except for within the Update GUI trigger function, when it's predetermined anyways.

gbakeman committed:
Small socket, power mode changes

- Socket Disconnect method now calls the Close_Socket method and raises the SocketDisconnected event in more cases
- System Resume event now tries to disconnect the UPS if it's still considered connected.

gbakeman committed:
Login, socket improvements

Nut_Socket.vb
- Removed Socket object and any relating code since TcpClient is already providing the functionality.
- Remove Disconnected event since this will be provided through exceptions or intentional commands.
- Fixed use of LOGIN protocol command
- Simplify Disconnect method

UPS_Device.vb
- Created dedicated Login method that interfaces with the socket layer.

WinNUT.vb
- Calls Login method only when a username is provided. Fixes forced/unintended logins.

gbakeman committed:
More smaller networking bugfixes

Nut_Socket.vb
- ConnectionStatus property checks for null client object before checking .Connected property
- TcpClient object is left null at construction, and only instantiated during a Connection subroutine.
- Disconnect subroutine doesn't explicitly check for connection status on the TcpClient object before running anymore. This allows it to dispose/close all objects regardless.

UPS_Device.vb
- Fixed polling interval not being applied to timer. Associated property is now read only, and the interval is set directly on the timer during construction.
- Fixed missing call to socket's Disconnect subroutine, and added a general exception catch-all.
- Removed Login code from Reconnect subroutine, leaving that to external code.
- Removed ReConnected event since it seems to duplicate the Connected event's calls.

WinNUT.vb
- Moved Login call to connection finalization subroutine.
- Added LogException call during protocol error handling.

kirsch33 commented:
i am happy to see this PR--i previously have been using a fork of this repo I made for myself which the only change being the `MaxRetry` variable within `UPS_Device.vb` vastly increased. that is what has resolved the connection issues for me.

related to https://github.com/nutdotnet/WinNUT-Client/issues/65, i believe it makes sense for this PR to also address a configurable option for the `MaxRetry` interval.

considering this software is otherwise perfect for me besides the connection issue, it would be greatly appreciated and thank you for all your work on this!

gbakeman commented:
Hi @kirsch33, thanks for your interest and sorry you had to fork just for that small change! 😄 To be honest, I wasn't planning on tackling #65 here since I'm helping another user who's experiencing a multitude of issues at the moment. Generally though, I think I'm leaning more towards indefinite/persistent connections instead of a set amount before giving up - let me know in #65 if you think it should be different.

kirsch33 commented:
@gbakeman that sounds good to me! and its no issue. this app is perfect otherwise, or else I wouldnt have went thru all that trouble lol so kudos to you and the other maintainers.

gbakeman committed:
Smaller socket updates

- Removed redundant IsConnected property in favor of ConnectionStatus
- Renamed Nut and Net version properties and converted them to auto-implemented properties.
- Wrapped version query statements in try-catch blocks in case servers that throw an error for these queries (Synology in particular) may not have further problems with other queries.
- Added support for Net version 1.3
